### PR TITLE
Feature/chapter user progress

### LINKED ISSUE #448: Chapter user progress component
## Motivations
Implement the view which represents the progress of a user on a chapter. ([design](https://france-ioi.github.io/algorea-designs/03.Activities_02.Participation_c.User-situation-(individual)))

## Subtasks
Display, in the chapter user progress tab of items Screenshot 2021-08-18 at 15 17 06 , a table with results of the new [groupParticipantProgress](https://france-ioi.github.io/algorea-devdoc/api/#operation/groupParticipantProgress) (service:
* as rows:
  * first row (on a darker background style?): the result of the item itself (you only have its name, and its results are in 'item' in the service response)
  * a row per child in 'children'  in the service response
* as columns:
  * Content: the item name (with a link)
  * Latest activity: in "YY-MM-DD HH:MM"
  * Time spent: displayed with `| duration`
  * `# subm.` : submissions (a number)
  * score: score/validated as a score ring

### PR DESCRIPTION
**Pull request by Iloveall**
Fixes #448

### CODE REVIEW

**Review by smadbe — CHANGES_REQUESTED** [review]
I get that error each time I got in the progress tab of this item, I suppose this is not a normal behavior.
Screenshot 2021-08-23 at 18 42 16

File: src/app/modules/item/http-services/get-participant-progress.service.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }),
+    submissions: D.number,
+    timeSpent: D.number,
+    type: D.literal('Chapter', 'Task', 'Course', 'Skill'),
+    validated: D.boolean,
+  })),
+  item: D.struct({
+    hintsRequested: D.number,
+    itemId: D.string,
+    latestActivityAt: dateDecoder,
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/http-services/get-participant-progress.service.ts:30]: Fixed

File: src/app/modules/item/http-services/get-participant-progress.service.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  children: D.array(D.struct({
+    currentUserPermissions: permissionsDecoder,
+    hintsRequested: D.number,
+    itemId: D.string,
+    latestActivityAt: D.nullable(dateDecoder),
+    noScore: D.boolean,
+    score: D.number,
+    string: D.struct({
+      languageTag: D.string,
+      title: D.string,
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/http-services/get-participant-progress.service.ts:20]: Fixed

File: src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.html [hunk 3]
```diff
@@ -1 +1,65 @@
-<p>chapter-user-progress works!</p>
+<ng-container *ngIf="state$ | async as state">
+  <div class="spinner" *ngIf="state.isFetching">
+    <p-progressSpinner [style]="{ width: '3rem', height: '3rem' }"></p-progressSpinner>
+  </div>
+
+  <p class="error" *ngIf="state.isError" i18n>Error while loading the user progress</p>
+
+  <div *ngIf="state.isReady">
+    <p-table class="alg-table alg-table chapter-user-progress" [columns]="columns" [value]="state.data" [loading]="state.isFetching">
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.html]: Fixed

File: src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.html [hunk 4]
```diff
@@ -1 +1,65 @@
...
+          let-rowData
+          let-columns="columns"
+      >
+        <tr>
+          <td *ngFor="let col of columns">
+            <ng-container [ngSwitch]="col.field">
+              <ng-container *ngSwitchCase="'title'">
+                <a
+                  class="alg-link"
+                  [routerLink]="rowData | rawItemLink"
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.html]: Fixed

**Review by smadbe — CHANGES_REQUESTED** [review]
👍 code ok.
It is a pity to have some titles truncated while the time spent, submission count and the score columns are too large. Could you fix that?
Also, left align the score as the rest.
Screenshot 2021-08-31 at 15 27 45

**Review by smadbe — CHANGES_REQUESTED** [review]
much better :-) great 👍 

Ah you haven't tested the user chapter view when observing a user :-( ? It seems it does not work.
Screenshot 2021-09-01 at 09 16 03

**Comment by ThomasJuster** [comment]
> much better :-) great 
> 
> Ah you haven't tested the user chapter view when observing a user :-( ? It seems it does not work.
> Screenshot 2021-09-01 at 09 16 03

I could not test it either, I tried to watch [this user](https://dev.algorea.org/en/#/activities/by-id/786619260741032385;path=4702,4102,902356740789159624,839024045585287566;attempId=0/details/progress/chapter) and go to a [chapter of one of his groups](https://dev.algorea.org/en/#/groups/by-id/31961217542260212;path=/details)
But the backend call responds `403: insufficient access rights`

**Review by ThomasJuster — COMMENTED** [review]
I have 2 small comments, after that it's good to go :+1:

File: src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.ts [hunk 5]
```diff
@@ -1,12 +1,118 @@
...
+    },
+    {
+      field: 'submissions',
+      header: '# subm.',
+    },
+    {
+      field: 'score',
+      header: 'Score',
+    }
+  ];
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.ts]: Fixed

File: src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.ts [hunk 6]
```diff
@@ -1,12 +1,118 @@
...
+
+  refresh(): void {
+    this.refresh$.next();
+  }
 
-  constructor() { }
+  onClick(rowData: RowData): void {
+    if (!this.itemData || this.itemData.item.id === rowData.id) return;
+    const parentAttemptId = this.itemData.currentResult?.attemptId;
+    if (!parentAttemptId) return; // unexpected: children have been loaded, so we are sure this item has an attempt
```
[the remark this answers is not in the document]
Comment (Iloveall) [reply src/app/modules/item/pages/chapter-user-progress/chapter-user-progress.component.ts]: Fixed

**Review by ThomasJuster — APPROVED** [review]

**Review by smadbe — APPROVED** [review]
👍

**Pull request merged**